Issues #6, #7 - Adding 0.9 capabilities, auto-retrieval of orgId. Started implementation of ServerCloneApi

trevorflanagan opened the pull request:
@andreaturli This is a new pull request that supersedes https://github.com/cloudsoft/amp-dimensiondata/pull/14 . This should make it easier to review. 

I will add further branches and pull requests for the other commits that we have made thus far.

andreaturli commented on src/main/java/org/jclouds/dimensiondata/cloudcontroller/filters/OrganisationIdFilter.java:
```diff
@@ -0,0 +1,55 @@
+         indexOfVersionParam = path.indexOf("/", indexOfApiBase + 4);
+      }
+      String substring = path.substring(0, indexOfVersionParam);
+               substring = substring + "/" + DimensionDataCloudControllerComputeServiceAdapter.ORG_ID;
```
I like the simplification introduced by `getAccountApi()` which allows ORG_ID discoverability.
tbh I don't think using it this way is the way to go, as `DimensionDataCloudControllerComputeServiceAdapter` init may happen too late. What if we inject the accountApi in this filter  instead, is it a circular dependency?

trevorflanagan replied:
Thanks for the feedback @andreaturli I will give the injection a try to see if it is possible. My thoughts would be that `DimensionDataCloudControllerComputeServiceAdapter ` would always be initialized, but you might know better.

On your comment below regarding making the project open-source, I would totally agree. We would be very keen to get to this stage.

trevorflanagan replied:
@andreaturli I have played around with injecting the `AccountAPI` into `OrganisationIdFilter` and can see no way to properly inject. Are you able to help me out? Maybe there is an example you can point me at?
I am not overly familiar with Guice but I do understand that I would need to make changes to the context so that I can bind the `AccountAPI` into `OrganisationIdFilter`. For all of this to work I will need to alter all existing mock tests using `BaseDimensionDataCloudControllerMockTest` as they will rely on having a an orgId supplied by the Account API.

andreaturli commented:
thanks @trevorflanagan et all, I like the direction of this PR, just a minor comment.

I think we should properly open-source it and donate this provider to jclouds project. I'd be more than happy to guide you guys through the process, opening a jira ticket, issue the first PR against jclouds/jclouds-labs and so on. This will give a better visibility to the DimensionData provider and will get help from the entire jclouds community instead of just me. wdyt?

andreaturli commented:
@trevorflanagan I'd probably do something similar to what I've added to your branch (see commit https://github.com/cloudsoft/amp-dimensiondata/pull/18/commits/37ca981f9ab2b851257cd705bd8e3e4c2e537163) 
notice I've not fixed the tests, but hopefully it gives you the idea. Feel free to amend the commit as you prefer.

trevorflanagan commented:
@andreaturli Thanks for helping me out there. The change makes sense to me. I have updated the code and fixed the tests. We will probably revisit how orgId is loaded since every request will require 2 individual requests. One for the org id and then one for the actual request.

andreaturli commented:
Cool.

@trevorflanagan notice because the `getMyAccount()` is @Memoized this will asked just once

andreaturli commented:
lgtm, merging
